File: frontend/src/types/outgoingDelivery.ts
export interface OutgoingItemOperation {
  id: number;
  itemId: number;
  operationId: number;
  value: string | null;
  done: boolean;
  factQty: number | null;
  executedBy: string | null;
  executedAt: string | null;
  operation: { id: number; code: string; name: string; unit: string | null; phase: string };
}

export interface OutgoingDeliveryItem {
  id: number;
  deliveryId: number;
  skuId: number | null;
  article: string;
  name: string | null;
  quantity: number;
  factQuantity: number | null;
  weight: string | null;
  volume: string | null;
  unitCost: string | null;
  totalCost: string | null;
  sku: { id: number; article: string; name: string } | null;
  operations: OutgoingItemOperation[];
}

export interface OutgoingDelivery {
  id: number;
  number: string;
  partnerId: number;
  warehouseCode: string | null;
  isCrossDock: boolean;
  sourceIncomingDeliveryId: number | null;
  status: string;
  shipDate: string | null;
  actualDate: string | null;
  comment: string | null;
  createdBy: string;
  createdAt: string;
  updatedAt: string;
  partner: { id: number; name: string };
  items: OutgoingDeliveryItem[];
}

export interface OutgoingDeliveriesResponse {
  data: OutgoingDelivery[];
  total: number;
}

export interface OutgoingDeliveryItemInput {
  id?: number;
  article: string;
  name?: string;
  quantity: number;
  weight?: number;
  volume?: number;
}

export interface OutgoingDeliveryFormData {
  partnerId: number;
  warehouseCode?: string;
  isCrossDock?: boolean;
  shipDate?: string;
  comment?: string;
  items: OutgoingDeliveryItemInput[];
}

export function outgoingTotal(delivery: OutgoingDelivery): number {
  return delivery.items.reduce((sum, i) => sum + (i.totalCost != null ? Number(i.totalCost) : 0), 0);
}

File: mobil/core/model/src/main/java/com/npp/tsd/core/model/OutgoingDelivery.kt
package com.npp.tsd.core.model

import kotlinx.serialization.Serializable

@Serializable
data class OutgoingItemOperation(
    val id: Int,
    val itemId: Int,
    val operationId: Int,
    val value: String? = null,
    val done: Boolean = false,
    val factQty: Int? = null,
    val executedBy: String? = null,
    val executedAt: String? = null,
    val operation: OperationDto,
)

@Serializable
data class OutgoingDeliveryItem(
    val id: Int,
    val deliveryId: Int,
    val skuId: Int? = null,
    val article: String,
    val name: String? = null,
    val quantity: Int,
    val factQuantity: Int? = null,
    val weight: String? = null,
    val volume: String? = null,
    val unitCost: String? = null,
    val totalCost: String? = null,
    val sku: SkuRef? = null,
    val operations: List<OutgoingItemOperation> = emptyList(),
)

@Serializable
data class OutgoingDelivery(
    val id: Int,
    val number: String,
    val partnerId: Int,
    val warehouseCode: String? = null,
    val isCrossDock: Boolean = false,
    val sourceIncomingDeliveryId: Int? = null,
    val status: String,
    val shipDate: String? = null,
    val actualDate: String? = null,
    val comment: String? = null,
    val createdBy: String,
    val createdAt: String,
    val updatedAt: String,
    val partner: PartnerRef,
    val items: List<OutgoingDeliveryItem> = emptyList(),
)

@Serializable
data class OutgoingDeliveriesResponse(
    val data: List<OutgoingDelivery>,
    val total: Int,
)

@Serializable
data class ShipOutgoingDeliveryItemBody(
    val itemId: Int,
    val factQuantity: Int,
)

@Serializable
data class ShipOutgoingDeliveryBody(
    val items: List<ShipOutgoingDeliveryItemBody>,
)
